### ISSUE TITLE
(Docker apple M1) Invalid mapping of address 0x40080f9000 in reserved address space below 0x400000000000

### ISSUE THREAD

**Issue by kamilcglr** [microsoft/mssql-docker#668]
Hello, 

After I pulled the mcr.microsoft.com/mssql/server:latest image, I have a weird error when I want to launch it.
How can I resolve that ? Thank you in advance.
```
docker run --platform linux/amd64 -e 'ACCEPT_EULA=Y' -e 'SA_PASSWORD=Passw0rd'  -v sql_server:/var/opt/mssql -p 1433:1433 -it mcr.microsoft.com/mssql/server:latest
SQL Server 2019 will run as non-root by default.
This container is running as user mssql.
To learn more visit https://go.microsoft.com/fwlink/?linkid=2099216.
/opt/mssql/bin/sqlservr: Invalid mapping of address 0x400976b000 in reserved address space below 0x400000000000. Possible causes:
1) the process (itself, or via a wrapper) starts-up its own running environment sets the stack size limit to unlimited via syscall setrlimit(2);
2) the process (itself, or via a wrapper) adjusts its own execution domain and flag the system its legacy personality via syscall personality(2);
3) sysadmin deliberately sets the system to run on legacy VA layout mode by adjusting a sysctl knob vm.legacy_va_layout.
```

It seems to be the same problem in this SO question : `https://stackoverflow.com/questions/65398641/docker-connect-sql-server-container-non-zero-code-1`

**Comment by RS-Roshi** [comment]
Any luck?

**Comment by kamilcglr** [comment]
> Any luck?

Unfortunately, no. 
But has suggested by u/slickvik4 in [reddit_comment](https://www.reddit.com/r/docker/comments/kelobr/help_building_sql_server_container_on_m1/ghcz3yf?utm_source=share&utm_medium=web2x&context=3) you can use azure sql edge image. This solution is not working at 100% I still have some issues with heavy request. I get the following error : "There is insufficient system memory in resource pool 'internal' to run this query".

**Comment by RS-Roshi** [comment]
I am also unable to connect to docker registry from my M1 Mac Air. I saw nobody is responding any solution to all these issue out there for new m1's

**Comment by daviduzumeri** [comment]
Also having this issue and it's basically the main thing blocking my org from being able to use M1s for development. I'm setting up a remote Ubuntu Docker host at home for this but it'd be really wonderful if this was fixable, even if just in QEMU x86 emulated mode in Docker (although an ARM64 native container would be incredible)

**Comment by kamilcglr** [comment]
It has been more than a month and no one has a solution. Maybe it will get better with the next tech preview.

**Comment by bradwallen** [comment]
Watching this space as well.

**Comment by rgioia** [comment]
any news?

**Comment by kamilcglr** [comment]
An engineer from Docker said : 
"The best fix for this would be for Microsoft to provide an arm64 image.
My guess is that this is a qemu emulation issue that won't be easy to debug or fix."

The best alternatives today are to use Azure SQL Edge or other database technologies such as mariadb or mysql which are multi arch.
[https://www.reddit.com/r/docker/comments/kelobr/help_building_sql_server_container_on_m1/)](discussion from reddit)

> Any news?

**Comment by daviduzumeri** [comment]
If Microsoft devotes resources to an ARM64 image, I will eat my hat :(

**Comment by richardksizemore** [comment]
Yes i went the Azure SQL Edge route and no issues for my purpose. https://hub.docker.com/_/microsoft-azure-sql-edge

**Comment by sshashka** [comment]
> Yes i went the Azure SQL Edge route and no issues for my purpose. https://hub.docker.com/_/microsoft-azure-sql-edge

could you please help me installing it? Because after this step: "sudo docker run --cap-add SYS_PTRACE -e 'ACCEPT_EULA=1' -e 'MSSQL_SA_PASSWORD=yourStrong(!)Password' -p 1433:1433 --name azuresqledge -d mcr.microsoft.com/azure-sql-edge" nothing happens

**Comment by rgioia** [comment]
I can install azure without any problems. My current problem is to be able to install sql-server to run a Ruby on Rails application with sql-server as adapter

**Comment by RS-Roshi** [comment]
> > Yes i went the Azure SQL Edge route and no issues for my purpose. https://hub.docker.com/_/microsoft-azure-sql-edge
> 
> could you please help me installing it? Because after this step: "sudo docker run --cap-add SYS_PTRACE -e 'ACCEPT_EULA=1' -e 'MSSQL_SA_PASSWORD=yourStrong(!)Password' -p 1433:1433 --name azuresqledge -d mcr.microsoft.com/azure-sql-edge" nothing happens


running this command
docker run --cap-add SYS_PTRACE -e 'ACCEPT_EULA=1' -e 'MSSQL_SA_PASSWORD=yourStrong(!)Password' -p 1433:1433 --name azuresqledge -d mcr.microsoft.com/azure-sql-edge

generates this result and it started on port 1433
344f7269f806c5db985c61ec78b1bf7e6398f8f398acc8ea694a741f0c3aace3

**Comment by chris-larsen** [comment]
Gave up on the Azure Edge image. Ran into the CLR error (I'd imagine quite a few DB's might depend on the CLR runtime). Saw Microsoft pulled CLR support for Azure Edge 4 years back due to a security risk and it never came back.

**Comment by SvyatovetsD** [comment]
Any progress on this? Azure SQL Edge works fine, but can't really use it due to lack of CLR support

**Comment by bradwallen** [comment]
> Gave up on the Azure Edge image. Ran into the CLR error (I'd imagine quite a few DB's might depend on the CLR runtime). Saw Microsoft pulled CLR support for Azure Edge 4 years back due to a security risk and it never came back. 

Same. I need CLR support.

**Comment by Gilbert1391** [comment]
Same issue here, is there any work around for this? I have an older MacBook where I can run the container, how can I connect to the db server from my M1? Both computers are connected to the same network. Any help would be much appreciated!

**Comment by chris-larsen** [comment]
> Same issue here, is there any work around for this? I have an older MacBook where I can run the container, how can I connect to the db server from my M1? Both computers are connected to the same network. Any help would be much appreciated!

There’s no fix ATM until Microsoft fixes the image on their end. Won’t work with dockers preview that includes emulation for m1. 

Best you can do is the azure edge image which is ARM compatible but you lose things like CLR runtime.

If you’re just doing simple stuff with small ish databases (I was learning with adventure works) you can pretty easily spin up a mssql server on the free google cloud tier and upload your database to a bucket there, whitelist the IP you’ll be connecting from and connect to it there. That’s what I ended up doing so I could continue my learning

**Comment by rgioia** [comment]
Any news?

**Comment by nljsystemJF** [comment]
> Yes i went the Azure SQL Edge route and no issues for my purpose. https://hub.docker.com/_/microsoft-azure-sql-edge
Azure SQL Edge works fine on my M1 Mac.
But unfortunately, sqlcmd tool is not available the ARM64 version of SQL Edge containers.

**Comment by aroon0808** [comment]
Any news?

**Comment by soccypowa** [comment]
I don't know if this helps this issue move at all, but I realized that building a custom image starting with Ubuntu 18.04 you will run into this problem as soon as you install sqlserver on it... Steps to reproduce:
`FROM --platform=amd64 ubuntu:18.04
RUN useradd -u 10001 mssql
RUN apt-get update && apt-get install -y wget software-properties-common apt-transport-https
RUN wget -qO- https://packages.microsoft.com/keys/microsoft.asc | apt-key add -
RUN add-apt-repository "$(wget -qO- https://packages.microsoft.com/config/ubuntu/18.04/mssql-server-2019.list)"
RUN apt-get update && apt-get install -y mssql-server
USER mssql
CMD /opt/mssql/bin/sqlservr`

**Comment by thisisthekap** [comment]
@soccypowa Does it work if you use the image based on Ubuntu 16.04 (e.g. the tag `2019-CU9-ubuntu-16.04`)?

**Comment by thisisthekap** [comment]
@LuisBosquez @twright-msft @amvin87 Are you able to give us any insight on possible workarounds or an ETA for a possible fix?

**Comment by soccypowa** [comment]
@thisisthekap Tried a few different starting points now, just for the "fun" of it, started with Ubuntu 16.04 and 18.08, tested both SQL 2107 and 2019, and to top it of I even built on CentOS 8 one last time... Still exits with the same error. I found a support article on MSFT about RHEL 7.4, it is the ball park of the same error, tried to, I am not very experienced with this, but to set the limits as pointed out in it, no success, [SQL Server terminates and generates a core dump on RHEL 7.4 when you try to run mssql-conf](https://docs.microsoft.com/en-US/troubleshoot/sql/linux/core-dump-rhel-7-4-run-mssql-conf). Even found this article that also is "kinda" the same [MSSQL DOCKER ON UNRAID](https://forums.unraid.net/topic/82110-mssql-docker-on-unraid/).

Must be something it does that either is not allowed by the tech preview itself or the fact that we are virtualizing amd64 hardware and that there is some limitation of what Apple thinks is ok.

**Comment by soccypowa** [comment]
There is a new RC released today, RC2, they have removed the preview banner at the bottom. Still though for our situation no cigar, tried both qemu and virtualization.framework, , official image as well as my custom build...
[https://docs.docker.com/docker-for-mac/apple-m1/](https://docs.docker.com/docker-for-mac/apple-m1/)

**Comment by kovalbogdan95** [comment]
I want to suggest a temporary workaround for this issue by using Azure SQL container.

Here my docker-compose.yml config example:
```yaml
  # Database instance
  mssql:
    image: mcr.microsoft.com/azure-sql-edge:latest
    container_name: mssql
    volumes:
      - events_mssql:/var/opt/mssql
      - ./setup.sql:/usr/config/setup.sql
    ports:
      - 1433:1433
    environment:
      - ACCEPT_EULA=Y
      - MSSQL_SA_PASSWORD=Passw@rd
      - MSSQL_PID=Developer
      - MSSQL_DB=events_service
      - MSSQL_USER=SA
      - MSSQL_PASSWORD=Passw@rd
      - MSSQL_DB_AUDIT_LOG=events_service_audit_log

volumes:
  events_mssql:
```
setup.sql file should be placed on the same level with the docker-compose.yml file
```sql
CREATE DATABASE $(MSSQL_DB);
CREATE DATABASE $(MSSQL_DB_AUDIT_LOG);
GO
USE $(MSSQL_DB);
GO
CREATE LOGIN $(MSSQL_USER) WITH PASSWORD = '$(MSSQL_PASSWORD)';
GO
CREATE USER $(MSSQL_USER) FOR LOGIN $(MSSQL_USER);
GO
ALTER SERVER ROLE sysadmin ADD MEMBER [$(MSSQL_USER)];
GO
```
Answer on StackOverflow: https://stackoverflow.com/questions/65398641/docker-connect-sql-server-container-non-zero-code-1/66919852#66919852

**Comment by thisisthekap** [comment]
@kovalbogdan95 Thanks to the insights,
But unfortunately, Azure SQL Edge does not support Full Text Search and a bunch of other more or less widely used features.

**Comment by kovalbogdan95** [comment]
@thisisthekap Sure, I'd rather run MSSQL container too, but for time being it's everything we got.

**Comment by thisisthekap** [comment]
@kovalbogdan95 Anonther workaround I could think of would be using a Ubuntu or Devian VM hosted e.g. in azure as Docker host. The local CI can easily be configured to reflect this change. A custom DNS entry could serve as hostname to access docker services.

**Comment by kovalbogdan95** [comment]
@thisisthekap Yes, it can solve the problem with DB, but if you want to use docker for local app development - it will cause a new problem where you can't mount the local folder to the remote container.
For me, it's critical, because I want to run both DB and app on my local machine.

I totally agree that remote DB is also a possible solution 😃

**Comment by thisisthekap** [comment]
@kovalbogdan95 Full agree.

**Comment by nonamez** [comment]
Also faced this problem...

**Comment by vKarter** [comment]
I have also the same issue on image 2019-latest

**Comment by tracker1** [comment]
@nljsystemJF late reply, but you could just use sqltools in the container...

**Comment by nikiask** [comment]
I also have this issue on an m1 arm mac. In my case I can use the azure sql edge in a docker container, but I still need mssqlodbc-17 to access mssql as a client/driver from another container (using docker compose). Even using rosetta to emulate x86 with docker crashes all the time, apparently due to qemu issues. I'm working on finding an odbc driver that works on arm, but I hope that mssqlodbc (and mssql in general) gets ported to arm.

@tracker1 I don't think sqltools comes with azure sql edge containers, which is the only version that is arm64.

**Comment by kovalbogdan95** [comment]
@nikiask I had a similar issue with my python app. I had to switch to a different client library - https://stackoverflow.com/questions/66896068/pyodbc-on-m1-macs/66919686#66919686
I switched pyodbc to pymssql (freetds based). I guess something can exist for your programming language too

**Comment by badsyntax** [comment]
Has MS actually confirmed this bug yet? It's a pretty major blocker for anyone trying to build .net applications with sql server on MacOS with the M1 chip. It would be great to get some feedback from MS.

**Comment by amvin87** [comment]
@SQLSourabh / @VasiyaKrishnan Can you guys take a look at this please.

**Comment by 2nal2** [comment]
any news?

**Comment by VasiyaKrishnan** [comment]
Work is currently in progress for Apple M1 (ARM64) support for the msodbcsql17 and mssql-tools (sqlcmd/bcp) packages. Will keep the thread updated when it is released.

**Comment by AnaghHegde** [comment]
@VasiyaKrishnan @amvin87 Is there any way to older version of SQL edge docker images ? Docker hub just shows the latest tag.

**Comment by dotariel** [comment]
Thanks for the update, @VasiyaKrishnan. Will this update also address the SQL Server 2017 Docker image being able to run on the M1?

**Comment by SQLSourabh** [comment]
> @VasiyaKrishnan @amvin87 Is there any way to older version of SQL edge docker images ? Docker hub just shows the latest tag.

Is there a reason why you are looking to run older version of SQL Edge image? For all purposes, you should be using the latest image.

**Comment by SQLSourabh** [comment]
> Thanks for the update, @VasiyaKrishnan. Will this update also address the SQL Server 2017 Docker image being able to run on the M1?

there are currently no plans to support SQL Server docker images on M1 platforms. You can run SQL Edge on M1. The work Siya highlighted is in the tools and drivers space, which will allow you to run SQL Tools on M1.

**Comment by Gilbert1391** [comment]
> > Thanks for the update, @VasiyaKrishnan. Will this update also address the SQL Server 2017 Docker image being able to run on the M1?
> 
> there are currently no plans to support SQL Server docker images on M1 platforms. You can run SQL Edge on M1. The work Siya highlighted is in the tools and drivers space, which will allow you to run SQL Tools on M1.

That's a shame, because there are some features such as Full Text Search that are unsupported for SQL Edge. I recently just ran into this problem, I was using the SQL Edge docker image as a workaround, but then hit a wall when I needed this FTS feature. I ended up setting a db in the cloud instead.

**Comment by dotariel** [comment]
> > Thanks for the update, @VasiyaKrishnan. Will this update also address the SQL Server 2017 Docker image being able to run on the M1?
> 
> there are currently no plans to support SQL Server docker images on M1 platforms. You can run SQL Edge on M1. The work Siya highlighted is in the tools and drivers space, which will allow you to run SQL Tools on M1.

This is unfortunate. I'm sure Azure SQL Edge is suitable for certain workloads, but many customers are still developing application that run on SQL Server 2017/19 that require certain capabilities that Azure SQL Edge does not support, and setting up a remote database is not practical for local development.

**Comment by aminta** [comment]
I'm following, SQL Server is needed also in my work...

**Comment by kovalbogdan95** [comment]
> I'm following, SQL Server is needed also in my work...

Same for me. Temporary using Azure SQL Edge

**Comment by vincevertulfo** [comment]
Any updates on this?

**Comment by gytrvi** [comment]
any updates?

**Comment by redaxmedia** [comment]
I need support for M1 aswell...

**Comment by mbrookson** [comment]
+1 for M1 support as well.

I imagine this will become a wider issue when the next generation of M1(X) machines are released this year as more people will be upgrading to the new architecture.

**Comment by robjwilkins** [comment]
😱 I have a new Mac with M1 chip and need to run MSSQL Server in docker, but get this error. Is it not possible?

**Comment by 2nal2** [comment]
@robjwilkins you can try https://hub.docker.com/_/microsoft-azure-sql-edge in the mean while sql server is realeased, it does not have all the functionality, but it works for me in development

**Comment by amvin87** [comment]
Thank you for all the comments. As of today, as indicated earlier by @SQLSourabh and @VasiyaKrishnan SQL Edge is supported on ARM. We are triaging the ARM support internally and will let you know on the next updates as soon as we have one.

**Comment by robjwilkins** [comment]
SQL Edge is not a viable alternative to the MSSQLServer image. It does not have `sqlcmd` in the package, therefore my Dockerfile will not work.

Seems premature closing this issue till the image does work with the M1 chip.

**Comment by thisisthekap** [comment]
As this issue is not solved, this issue definitely should be reopened. The right time to close this one would imho be the time when a fixed version makes it to the docker hub.

**Comment by patrickue** [comment]
> As this issue is not solved, this issue definitely should be reopened. 

If Microsoft decided to NOT solve this issue it makes sense for them to mark this issue CLOSED. You might disagree with that decision, but why should they keep this issue open if they don't plan to ever solve it. Marking it **closed** is a clear statement.

**Comment by thisisthekap** [comment]
@patrickue A clear statement would be "a clear statement". Just closing it with some vague announcement is imho not a clear statement, that Microsoft is definitely not solving this issue.

**Comment by SQLSourabh** [comment]
Hi, 

This issue was closed as there are no new updates, other than what has already been shared. If and when we have new updates, we will reopen this thread to share the updates.
<< Recap >> 
 At present, there is NO support for SQL Server on M1 or ARM processors. We are evaluating this, however we do not have an ETA of when this will be available. In the mean time, the supported mechanism are to either use Azure SQL Edge on MI/ARM, or use a remote SQL Server instance or Azure SQL. 
<< Recap >>

**Comment by chris-larsen** [comment]
Yikes, feel like this is really demonstrating the fragmentation of Microsoft's various compartments in their business and how distant each unit operates from another. Lack of vision from the top down as Microsoft is pretty on board with ARM on their own hardware and are working pretty hard on good emulation to entice people to make the switch. Yet over in this field of grass it appears this unit of Microsoft doesn't seem to recognize that ARM is going to be our next decades performant chip of choice and we will see more and more non-Mac ARM equipped computers up ahead (just follow the development of Qualcomm). 

I say keep the issue open and on the radar, consult some architects on the future of computing and try to strategize around that. There's no shortage of comments here and on many other issues where the azure sql edge image falls impossibly short of the sql server image, so if you guys insist on keeping this closed at least think about bringing the azure stuff in parity with what exists in sql server. Can almost feel the fragmentation of this sect of Microsoft too, one day we can't tell if you guys are betting on azure solutions, azure data studio, etc. and another we see announcements on new things coming to SSMS. Hate to be critical but we can really feel when it starts feeling like Microsoft is either spreading themselves too thin and/or are simply trying to dictate the direction of where certain technologies are headed rather than building technologies around the needs people have.

**Comment by tracker1** [comment]
> SQL Edge is not a viable alternative to the MSSQLServer image. It does not have
> `sqlcmd` in the package, therefore my Dockerfile will not work.

Can SQLCMD from the mssqlserver image connect to a remote instance of sql edge?  I've been running some publish scripts via docker[1] for a while, but usually targeting a different (non-local) instance.  Just a thought/suggestion.  In my case, I'm running some initial scripts ported from an SQL project in VS, then using versioned migration scripts.. this allow(ed) me to run my initial migrations, or run more complex SQL file scripts without issues (the setvar stuff namely).

Not sure if the client can run, even if the server itself has issues.

[1] https://github.com/tracker1/node-ms-sqlcmd

**Comment by jakubjanis** [comment]
I've just filed a suggestion on this topic at MS's feedback site.

It can be found here:
https://feedback.azure.com/d365community/idea/949650e5-3248-ec11-a819-0022484bf651

**Comment by tracker1** [comment]
Has anyone tried running the `sqlcmd` executable from the mssql image in Docker for MacOS in an M1?  Not the server, just `sqlcmd`.  I now have an M1, but mostly remote developing via my desktop... may give this a try at some point... possibly publish a container based on alpine meant for just sqlcmd usage.  I've had pretty good luck using SQL Edge in docker so far, but hadn't tried for the use case(s) above.

**Comment by Meligy** [comment]
Has anyone who uses Full-Text Search been able to get this sorted somehow?

Does Edge actually work at all? How would you add Full Text Search and sqlcmd to it?

Thanks heaps.

**Comment by thisisthekap** [comment]
I ende up using a Ubuntu VM in Azure with Docker installed. I added the remote docker instance as docker context on my local machine.

**Comment by thisisthekap** [comment]
@Meligy Is this a feasible option to you?

**Comment by Meligy** [comment]
> @Meligy Is this a feasible option to you?

No. I ended up using x64 Linux in UTM, and running docker in it.

https://twitter.com/Meligy/status/1473680420596338694

**Comment by jsbaguette** [comment]
I have to say I am surprise that a company like microsoft can handle that :s Sad to read the thread.

**Comment by bw-flagship** [comment]
I am very disappointed that Microsoft seems to have decided not to enable the SQL Server Docker image for Mac in the future (all Macs will use M1).

**Comment by ctholho** [comment]
I know, who am I to complain but the the slow m1 support is really flushing the goodwill ms created in the last few years with me down the drain.

**Comment by strangesast** [comment]
there is a [golang port of sqlcmd](https://github.com/microsoft/go-sqlcmd) that is supported on m1, easily run from a golang docker image
```
go get github.com/microsoft/go-sqlcmd/cmd/sqlcmd@v0.2.0
```

**Comment by ash3rr** [comment]
Using the latest image for Azure SQL Edge is giving me the same error. 

`docker run --cap-add SYS_PTRACE -e 'ACCEPT_EULA=1' -e 'MSSQL_SA_PASSWORD=Password1234' -p 1433:1433 --name azuresqledge --platform linux/amd64 -d mcr.microsoft.com/azure-sql-edge`

Azure SQL Edge will run as non-root by default.
This container is running as user mssql.
To learn more visit https://go.microsoft.com/fwlink/?linkid=2140520.
/opt/mssql/bin/sqlservr: Invalid mapping of address 0x40092ba000 in reserved address space below 0x400000000000. Possible causes:
1) the process (itself, or via a wrapper) starts-up its own running environment sets the stack size limit to unlimited via syscall setrlimit(2);
2) the process (itself, or via a wrapper) adjusts its own execution domain and flag the system its legacy personality via syscall personality(2);
3) sysadmin deliberately sets the system to run on legacy VA layout mode by adjusting a sysctl knob vm.legacy_va_layout.

**Comment by htonheim** [comment]
> Using the latest image for Azure SQL Edge is giving me the same error.
> 
> `docker run --cap-add SYS_PTRACE -e 'ACCEPT_EULA=1' -e 'MSSQL_SA_PASSWORD=Password1234' -p 1433:1433 --name azuresqledge --platform linux/amd64 -d mcr.microsoft.com/azure-sql-edge`
> 
> Azure SQL Edge will run as non-root by default. This container is running as user mssql. To learn more visit https://go.microsoft.com/fwlink/?linkid=2140520. /opt/mssql/bin/sqlservr: Invalid mapping of address 0x40092ba000 in reserved address space below 0x400000000000. Possible causes:

Azure SQL Edge has a native ARM image. Omit the `--platform linux/amd64` and it should work. You might have to delete the image first as well. At least I had to. Delete the container and image and then run it again without the platform option and it should work and run natively instead of through emulation.

**Comment by kitloong** [comment]
> Unsupported features
> 
> * CLR assemblies, and related DDL commands and Transact-SQL functions, catalog views, and dynamic management views.
> * CLR-dependent T-SQL functions, such as ASSEMBLYPROPERTY, FORMAT, PARSE, and TRY_PARSE.
> * CLR-dependent date and time catalog views, functions, and query clauses.

:(

https://docs.microsoft.com/en-us/azure/azure-sql-edge/features#unsupported-features

**Comment by Meligy** [comment]
> > Unsupported features
> > 
> > * CLR assemblies, and related DDL commands and Transact-SQL functions, catalog views, and dynamic management views.
> > * CLR-dependent T-SQL functions, such as ASSEMBLYPROPERTY, FORMAT, PARSE, and TRY_PARSE.
> > * CLR-dependent date and time catalog views, functions, and query clauses.
> 
> :(
> 
> https://docs.microsoft.com/en-us/azure/azure-sql-edge/features#unsupported-features

Also, the deal breaker in my case, Full-text search:

> Full-text indexes and search, and related DDL commands and Transact-SQL functions, catalog views, and dynamic management views.

**Comment by kitloong** [comment]
> > @Meligy Is this a feasible option to you?
> 
> No. I ended up using x64 Linux in UTM, and running docker in it.
> 
> https://twitter.com/Meligy/status/1473680420596338694

Hi @Meligy, it would be very nice if you could share how you manage to install Debian AMD64 using UTM!

**Comment by kitloong** [comment]
> > @Meligy Is this a feasible option to you?
> 
> 
> No. I ended up using x64 Linux in UTM, and running docker in it.
> https://twitter.com/Meligy/status/1473680420596338694 

Thank you @Meligy! I was finally able to run SQL Server 2019 in UTM.

I have written an article to help those who are looking for the same solution:
https://kitloong.medium.com/how-to-run-sql-server-2019-on-macbook-pro-m1-d1448525f805

**Comment by baluubas** [comment]
I have been using [Colima](https://github.com/abiosoft/colima) with the arch-flag `colima start --arch amd`. It's not fast but works.

**Comment by vvanghelue** [comment]
Please fix it

**Comment by aharpervc** [comment]
For those of you using colima on an M1 Macbook, what's your exact process? Here's what I'm doing, but I'm getting the same error as @kamilcglr 

1. `brew install colima`
2. `colima start --arch x86_64`
  a. this is what's listed with `colima help start`, but I also tried `--arch amd`
3. `colima status` shows "colima is running", "runtime: docker", "arch: aarch64"
  a. I expected it to say "arch: x86_64", but maybe it means the host architecture?
4. `docker info` shows the colima context
5. `docker run --platform linux/amd64 --rm mcr.microsoft.com/mssql/server:2019-latest` shows the same error:

```
SQL Server 2019 will run as non-root by default.
This container is running as user mssql.
To learn more visit https://go.microsoft.com/fwlink/?linkid=2099216.
/opt/mssql/bin/sqlservr: Invalid mapping of address 0x4003817000 in reserved address space below 0x400000000000. Possible causes:
1) the process (itself, or via a wrapper) starts-up its own running environment sets the stack size limit to unlimited via syscall setrlimit(2);
2) the process (itself, or via a wrapper) adjusts its own execution domain and flag the system its legacy personality via syscall personality(2);
3) sysadmin deliberately sets the system to run on legacy VA layout mode by adjusting a sysctl knob vm.legacy_va_layout.
```

**Comment by Meligy** [comment]
@aharpervc did you try `docker context list` and switching the context to `colima` if there's a context for it?

Based on [colima's docs][doc]. I'll probably try this too as it's slightly nicer than my UTM flow working with `DOCKER_SOCK` environment variable.

[doc]: https://github.com/abiosoft/colima#:~:text=Can%20it%20run%20alongside%20Docker%20for%20Mac%3F

**Comment by Meligy** [comment]
I also tried `colima` and it didn't work with `x86_64` or `amd`.

Back to UTM for now.

**Comment by baluubas** [comment]
> I also tried `colima` and it didn't work with `x86_64` or `amd`.
> 
> Back to UTM for now.

Try passing the --profile flag when you start the VM. I think a profile is a way to have multiple named VMs. You'll  also get a different docker context corresponding to each profile

colima start --profile amd --arch amd

**Comment by aharpervc** [comment]
@meligy the thing that tripped me up is that if you run `colima start` with no parameters you'll get a VM that matches the host architecture, eg aarch64. `colima status` will show this as well. To **switch** architectures after you've already run `colima start`, you have to run `colima delete` first. Then you can do `colima start --arch amd` to get a VM with the more useful x86_64 architecture (again, `colima status` to confirm).

Using the profile option as @baluubas said seems like a good way to keep track as well.

**Comment by Meligy** [comment]
Update: colima actually worked for me, even without a non default profile (I did colima delete a few times while testing though). There was a problem in my init scripts and I confused the error message with the one related to this issue.

Not I don't have to keep the UTM window open in another screen all the time. Thanks for both tips though.

**Comment by B4nan** [comment]
After few hours of struggling, I managed to run the edge image on M1. I had several issues, the last one was I used tmpfs volumes - and even after removing the configuration from docker-compose file, they stayed - I had to wipe all the volumes and create everything from scratch. 

Here is what worked for me:

```yaml
  mssql:
    image: mcr.microsoft.com/azure-sql-edge
    ports:
      - "1433:1433"
    volumes:
      - mssql:/var/opt/mssql/data
    user: root
    environment:
      SA_PASSWORD: Root.Root
      ACCEPT_EULA: Y

volumes:
  mssql:
```

And I had to do:

```bash
# turn of all containers
docker-compose down --remove-orphans 
# remove all images
docker system prune -a 
# remove all volumes
docker volume rm $(docker volume ls -q)
# run them again
docker-compose up -d
```

(note that some of those commands might be a bit destructive, if you don't want to loose data from your docker volumes, you should probably just remove the volume connected to the mssql image)

**Comment by dkgndianko** [comment]
I had the same issue with my M1 pro.
Tried a first option with these steps
1. provisioning a new instance with `colima start`
2. Switch docker to use the newly created VM with `docker context use <name_of_newly_created_vm>`
3. run my docker command (`docker run` or `docker compose`)

But it didn't work.
So I tried another approach that is very straightforward with `limactl`:
1. Run the command `limactl start default` (you can choose a different name) and edit ressources (cpu, memory and disk)
2. Use `nerdctl` instead of `docker` to run containers. In my case it was `lima nerdctl compose up`.

And actually it works! 💪🏽 
I have only some issues to fix with my mounted data. It was mounted as read-only. And I guess I can fix that
```logs
db_1      |sqlservr: Unable to open /var/opt/mssql/.system/instance_id: File: pal.cpp:566 [Status: 0xC0000022 Access Denied errno = 0x1E(30) Read-only file system]
db_1      |/opt/mssql/bin/sqlservr: PAL initialization failed. Error: 101
```

For those interested in my config, here it is:
**docker-compose.yml**
```yaml
version: "3.1"
services:
  db:
    image: mcr.microsoft.com/mssql/server:2019-latest
    user: root
    environment:
      ACCEPT_EULA: Y
      SA_PASSWORD: "H1ddenMuus"
      MSSQL_PID: Developer  # (Developer, Express, Standard, Enterprise, EnterpriseCore
#      MSSQL_DATA_DIR: /var/opt/mssql
    entrypoint: /var/cidemia/scripts/docker_entrypoint.sh
    command:
      - /opt/mssql/bin/sqlservr
    ports:
      - "1433:1433"
    volumes:
      - ./databases/scripts:/var/cidemia/scripts
      - ./databases/data:/var/opt/mssql

  adminer:
    image: adminer
#    restart: always
    depends_on:
      - db
    ports:
      - "8080:8080"
```

**./databases/scripts/docker_entrypoint.sh**
```shell
#!/bin/bash
#start SQL Server and import the data
/var/cidemia/scripts/import-data.sh & exec "$@"
```

**./databases/scripts/import-data.sh**
```shell
#run the setup script to create the DB and the schema in the DB
#do this in a loop because the timing for when the SQL instance is ready is indeterminate
fileRoot=/var/cidemia/scripts
fileName=$fileRoot/create_schemas.sql
for i in {1..50};
do
    /opt/mssql-tools/bin/sqlcmd -S localhost -U sa -P "${SA_PASSWORD:-'H1ddenMuus'}" -d master -i $fileName
    if [ $? -eq 0 ]
    then
        echo "$fileName completed"
        break
    else
        echo "not ready yet..."
        sleep 1
    fi
done
```
**./databases/scripts/create_schemas.sql** where I put my schema creation
```sql
CREATE DATABASE cidemiaTest;
GO
USE cidemiaTest;
GO

CREATE SCHEMA portal;
GO
CREATE SCHEMA statistics;
GO
CREATE SCHEMA erp;
GO
```

**Comment by tracker1** [comment]
@dkgndianko - I would strongly recommend against using volumes to host paths for databases in Docker for Windows and Mac... the communication is *significantly* slower... you'd be much better off creating/using a docker volume in the docker environment itself for such things.  You can still copy in/out of that volume as needed.

**Comment by dkgndianko** [comment]
> @dkgndianko - I would strongly recommend against using volumes to host paths for databases in Docker for Windows and Mac... the communication is _significantly_ slower... you'd be much better off creating/using a docker volume in the docker environment itself for such things. You can still copy in/out of that volume as needed.
> 
> (Exception: if using Docker for Windows via WSL2, then a wsl2 volume path is fine)

---
Hi @tracker1 . Thank you for the recommendation.
Could you please develop more on the subject or maybe point me to useful reading to understand it more the pro and cons.
Also based to the docker compose file I shared above, how would you modify it to comply with that.

 ---
**[Precision]**
I no longer use Docker for it, but instead use *nerdctl* with *lima*. Do you think that the same point will apply there ?

---
Very interesting subject to develop and discuss.
Thank you again !

**Comment by tracker1** [comment]
@dkgndianko for the compose file, you can add 

```yaml
...
volumes:
  sqldata:
services:
  db:
    ...
    volumes:
      - sqldata:/var/opt/mssql
```

You create a named volume above or below services, defaults are usually fine.  inside the service definition, you reference that volume by name for the outside portion of your data.

When using `docker-compose down`, the volumes will remain even if the services are stopped... if you want to remove volume data, you can use `docker-compose down -v` for destructive stop... I usually use `-t 1` as well to force a faster shutdown in those cases.

You can still use docker to copy files in/out of that named volume (there's a prefix added to docker-compose volums/containers, the directory name by default.

**Comment by konstantinos-kafkas-cko** [comment]
Is there a plan to solve this properly?

**Comment by SuperFola** [comment]
As stated in this comment from this discussion, not really.

> Thank you for all the comments. As of today, as indicated earlier by @SQLSourabh and @VasiyaKrishnan SQL Edge is supported on ARM. We are triaging the ARM support internally and will let you know on the next updates as soon as we have one.

**Comment by JohannesHoppe** [comment]
+1

**Comment by JohannesHoppe** [comment]
I tried all instructions from @dkgndianko and I always end up with: 

```
db_1 |exec /var/cidemia/scripts/docker_entrypoint.sh: exec format error
```

This seems to indicate that lima can't run the entryfile at all. Do any of you have a solution for this? I've been trying for ages to get the SQL server to work locally on my M1 somehow.... Please help me!!! 🙏🙏🙏

**Comment by dkgndianko** [comment]
> I tried all instructions from @dkgndianko and I always end up with:
> 
> ```
> db_1 |exec /var/cidemia/scripts/docker_entrypoint.sh: exec format error
> ```
> 
> This seems to indicate that lima can't run the entryfile at all. Do any of you have a solution for this? I've been trying for ages to get the SQL server to work locally on my M1 somehow.... Please help me!!! 🙏🙏🙏

@JohannesHoppe I had this in the past, but do not remember how I solved it.
You may try to replace the Shebang (`#!/bin/bash`) by other shell interpreters (e.g `#!/bin/sh`, `#!/bin/zsh`, etc.)

**Comment by JohannesHoppe** [comment]
Thanks for the quick reply and your support here!  🙏 

I tried colima first, didn't worked, tried pure lime... Now I restarted my machine and followed the colima instructions again from scratch. Now it works! Why?? 🙈

Hint: `limactl edit` > `Location: "~"` > `writable: true` is still active. I don't know if this has any effect.

## For all the other suffering developers:
```
brew install colima
colima start --arch amd

colima status
# must show x86_64, otherwise 
colima delete
colima start --arch amd

docker-compose up
```

Here is my old `docker-compose.yml` v2 with server:2017-latest, just as a reference:

```
version: '2.1'

services:
  db:
    image: mcr.microsoft.com/mssql/server:2017-latest
    environment:
      - ACCEPT_EULA=Y
      - SA_PASSWORD=xxx
      - MSSQL_PID=Developer
      - MSSQL_LCID=1033
      - MSSQL_COLLATION=SQL_Latin1_General_CP1_CI_AS
      - MSSQL_MEMORY_LIMIT_MB=8192
      - MSSQL_AGENT_ENABLED=true
      - TZ=Europe/Berlin
    ports:
      - 1433:1433
      - 49200:1433
    volumes:
      - mssql-volume:/var/opt/mssql
      - /var/folders/:/var/folders/

volumes:
  mssql-volume:
```

**Comment by tracker1** [comment]
@githanium While that will work for many, especially if you aren't using advanced features, it's not really a full 1:1 to The full version(s) of SQL Server.

**Comment by thisisthekap** [comment]
Docker is going to add support without the need for QEMU: https://github.com/docker/roadmap/issues/384

**Comment by bastianpsycho1** [comment]
Hello. Any news :(¿?

**Comment by januszm** [comment]
`azure-sql-edge` image works well on M1 Mac, just one note about connecting to the container from the same Mac:

* install `sqlcmd` from https://docs.microsoft.com/en-us/sql/connect/odbc/linux-mac/install-microsoft-odbc-driver-sql-server-macos
* run the server `docker run -d --name mssql-server --platform linux/arm64/v8 -e ACCEPT_EULA=Y -e SA_PASSWORD=SECRET -p 1433:1433 mcr.microsoft.com/azure-sql-edge`
* connect with `-C` option as the docker container uses a self-signed certificate `sqlcmd -U sa -P SECRET -S 127.0.0.1,1433 -C`

**Comment by JohannesHoppe** [comment]
`azure-sql-edge` does not support all commands of the full server. This is not a solution for us.

**Comment by Will-Bill** [comment]
macOS Ventura will support Rossetta 2 for VMs. E.G we will be able to install Run x86_64 Apps On ARM Linux W/ Rosetta - which I guess means we will get docker support using Rossetta for MS SQL?

**Comment by jafetbjarkar** [comment]
Has anyone tried using the mssql docker image on the Ventura beta?

**Comment by AdamVyborny** [comment]
@jafetbjarkar yea, still doesn't work.

**Comment by Will-Bill** [comment]
I think the thread to follow is https://github.com/docker/roadmap/issues/384

**Comment by A-Programmer** [comment]
It might be very dummy mistake but I am going to tell, it might be useful for someone :
I was setting a simple password for SA user, I was using 123 as password and I didnt read the errors completely, I was thinking that because of M1 I am getting this error, suddenly I remembered that I can not set a simple password, I changed the password to strong one and it running now.

**Comment by jclausen** [comment]
I don't understand why this issue was closed.  Since M1 is now the only available processor arch for MacBook Pro machines, as developers upgrade their hardware, they lose the ability to use the SQL Server image.   

The Azure Edge image works for basic functionality against SQL server, but does not include full text, spatial functions, nor simple conversion functions like `cast( myColumnName as "hierarchyId")`.   

With the availability of Docker BuildX, I fail to see why building to support arm64 is problematic.   It seems to  me like this issue should be opened and fast-tracked.

**Comment by tracker1** [comment]
@jclausen If I were to guess... first, SQL Server and Azure SQL Edge are somewhat different beasts and the missing functionality may be more closely tied to x86... or just difficult to make work.  Beyond this, the underlying error may be because of Apple's and/or Docker's implementation... right now, from what I understand, Docker is using QEMU and it's less than perfect... there is work either in progress, or upcoming to get Docker Desktop on MacOS using Rosetta on M1 based macs, and this may well resolve the issue at hand.

I, personally wouldn't expect MS to address this anytime really soon as it's a combination of difficult to track down, difficult to diagnose and a relatively small number of users experiencing the issues that need missing functionality.

It may be more prudent to request the missing functionality (by feature) for the Azure SQL Edge image(s) than to resolve this issue.

**Comment by jesperkarlsson42** [comment]
+1 still doesn't work

**Comment by jclausen** [comment]
Suggest weighing in on this issue, which is still open.  https://github.com/microsoft/mssql-docker/issues/734

**Comment by Danielliudov** [comment]
Still doesn't work, or I am missing something why is this closed?

**Comment by tracker1** [comment]
https://github.com/docker/roadmap/issues/384

**Comment by ahmedmshaker** [comment]
> `azure-sql-edge` image works well on M1 Mac, at least for simple use cases, just one note about connecting to the container from the same Mac:
> 
> * install `sqlcmd` from https://docs.microsoft.com/en-us/sql/connect/odbc/linux-mac/install-microsoft-odbc-driver-sql-server-macos
> * run the server `docker run -d --name mssql-server --platform linux/arm64/v8 -e ACCEPT_EULA=Y -e SA_PASSWORD=SECRET -p 1433:1433 mcr.microsoft.com/azure-sql-edge`
> * connect with `-C` option as the docker container uses a self-signed certificate `sqlcmd -U sa -P SECRET -S 127.0.0.1,1433 -C`

this is the only solution that works with me thanks

**Comment by 4n4n4s** [comment]
why did this get closed? it seems it's still not working on arm mac?

**Comment by thisisthekap** [comment]
@4n4n4s You may try to use Rosetta 2 to run MSSQL Server on your arm mac (https://github.com/microsoft/mssql-docker/issues/734#issuecomment-1333696500). Works quite good for me. Way faster compared to QEMU emulation.

**Comment by Meligy** [comment]
> @4n4n4s You may try to use Rosetta 2 to run MSSQL Server on your arm mac ([#734 (comment)](https://github.com/microsoft/mssql-docker/issues/734#issuecomment-1333696500)). Works quite good for me. Way faster compared to QEMU emulation.

Just to be clear, this script is using Colima (like: a VM), not Docker Desktop for Mac.

**Comment by thisisthekap** [comment]
@Meligy Yes, it uses Lima as vm, and Colima as container runtime. This is somewhat similar to Docker on mac, which as well starts a VM.

Lima and Colima are used because Docker is not (yet) supporting the use of Rosetta 2 within the VM.

**Comment by invaderb** [comment]
Can confirm using lima and colima does resolve the issue with running MSSQL on an M1 mac

**Comment by mikart143** [comment]
I can confirm that using docker 4.16 with enabled rosetta enable to run mssql container with pretty decent speed. Check this issue for more info: https://github.com/docker/roadmap/issues/384.
<img width="979" alt="IMG_2597" src="https://user-images.githubusercontent.com/4084564/212267007-0174af74-40d2-4739-880a-f0bb221a390e.png">

**Comment by thevangelist** [comment]
I think the best options for Mac are:

## 1) Run AzureSQLEdge base image. 

ARM64 version of AzureSQLEdge doesn't include sqlcmd, so you need to enter database contents in some other way. AzureSQLEdge is probably best for small DBs.

## 2) Emulate another OS with UTM.

For example, I am using to run x86 architecture Docker images. Specific ports are configured open between VM and host. It's not an ideal solution, but at least you don't need ARM images at all. I am connected to the 1433 with my host's database tool and using the x86 image from Debian11 Docker!

Steps to make it happen:

1) Download [GetUTM](https://mac.getutm.app/)
2) Download [Debian 11](https://cdimage.debian.org/debian-cd/current/amd64/iso-cd/)
3) Install Debian. For example look at this [Youtube vid](https://www.youtube.com/watch?v=yCTx-Xg8uuQ)
4) Configure Debian in UTM to have open ports for your needs (for example 2375 > 2375 for Docker Daemon)
5) Install Docker inside Debian

Optional:

* Change DOCKER_HOST variable in your host ~/.zshrc (or exported env variable) to command your newly-installed Debian 11 Docker on UTM.

Note... This is somewhat of a tedious process but once it is done, it's pretty easy to use.

Big thanks to: [Maarten Merken: Docker Express: Running a Local SQL Server on Your M1 Mac](https://medium.com/geekculture/docker-express-running-a-local-sql-server-on-your-m1-mac-8bbc22c49dc9)

**Comment by JamesHisTheKing** [comment]
the above worked for me

**Comment by JamesHisTheKing** [comment]
> ## 3) Use Docker 4.16 and activate Rosetta emulation (beta feature)
> Activating this beta feature on an M1 Mac (tested: M1 Pro, M1 Max) makes the image (Microsoft SQL Server 2017 (RTM-CU31) (KB5016884) - 14.0.3456.2 (X64)) start without problems:
> 
> ![Bildschirm_foto 2023-02-01 um 15 05 41](https://user-images.githubusercontent.com/341793/216079489-a0247dbc-f981-4913-9674-7c382c6a93a1.png)

this works for me Thanks gr8bit

**Comment by daniel-mulroy-cko** [comment]
Any idea why I can't see this setting on the same version?
![image](https://user-images.githubusercontent.com/110408247/218975090-0e4f04e4-2142-4cf4-bee2-4f0d10c4c457.png)

**Comment by gr8bit** [comment]
Not an M1 Mac? ;) looks like you posted on my Intel MBP.

**Comment by daniel-mulroy-cko** [comment]
Definitely an M1 mac 😅 

<img width="201" alt="image" src="https://user-images.githubusercontent.com/110408247/218977140-4aa49a65-1254-43cd-b359-ff3ddf496894.png">

**Comment by thisisthekap** [comment]
@daniel-mulroy-cko You need to update to macOS Ventura.

**Comment by daniel-mulroy-cko** [comment]
@thisisthecap thank you very much

**Comment by thevangelist** [comment]
I can confirm this actually works. I am running a `mcr.microsoft.com/mssql/server:2022-latest`

- Install Ventura
- Upgrade Docker to latest
  - Docker Settings > General: `[X] Use virtualization framework` and
  - Docker Settings > Features in Development: `[X] Use Rosetta...`
- Use `platform: linux/amd64` in Docker-compose or `--platform linux/amd6` when running containers from command line

E: Update 2026/1
- 2025-latest often crashes with AVX error on macOS Tahoe (Docker + Rosetta)
- Always use `-e "MSSQL_PID=Developer"` for dev license
- Add a volume for data e.g. `-v mssql-data:/var/opt/mssql` to survive restarts

**Comment by dirkpostma** [comment]
For anyone looking for the `Use Rosetta for x86/amd64 emulation on Apple Silicon` setting and not seeing it:

- make sure you have Docker 4.16 or higher
- make sure you have MacOS 13 Ventura (or higher)

**Comment by caematos** [comment]
> For anyone looking for the `Use Rosetta for x86/amd64 emulation on Apple Silicon` setting and not seeing it:
> 
> * make sure you have Docker 4.16 or higher
> * make sure you have MacOS 13 Ventura (or higher)

Can that be done with colima?

**Comment by JonMoore46** [comment]
> ## 3) Use Docker 4.16 and activate Rosetta emulation (beta feature)
> Activating this beta feature on an M1 Mac (tested: M1 Pro, M1 Max) makes the image (Microsoft SQL Server 2017 (RTM-CU31) (KB5016884) - 14.0.3456.2 (X64)) start without problems:
> 
> ![Bildschirm_foto 2023-02-01 um 15 05 41](https://user-images.githubusercontent.com/341793/216079489-a0247dbc-f981-4913-9674-7c382c6a93a1.png)

This worked for me as well 👍

**Comment by gohokies** [comment]
I run both mcr.microsoft.com/mssql/server:2022-latest and mcr.microsoft.com/mssql/server:2019-latest images successfully on M1 macbook with "Use Rosetta for x86/amd64 emulation on Apple Silicon" feature enabled in Docker Desktop V4.19.0.

**Comment by nagyDarius** [comment]
Hi, I'm trying to run this on the latest versions:
Docker v4.19.0
MacOS Ventura 13.3.1

But enabling the rosetta option results in this log when running the container.

2023-05-17 15:33:31 -- Not first container startup --
2023-05-17 15:33:33 rosetta error: unexpected BrkKind 17 on SIGTRAP

Any ideas why?

**Comment by AlexcastroDev** [comment]
This work for me:

```yaml
version: '3.8'
services:
  mssql:
    image: mcr.microsoft.com/azure-sql-edge
    environment:
      - ACCEPT_EULA=Y
      - SA_PASSWORD=Admin123@@
      - MSSQL_PID=Developer
    ports:
      - 1433:1433
    volumes:
      - ./data:/var/opt/mssql
```

**Comment by Rajesh2015** [comment]
Try this for Mac m1 docker run -e "ACCEPT_EULA=1" -e "MSSQL_SA_PASSWORD=reallyStrongPwd123" -e "MSSQL_PID=Developer" -e "MSSQL_USER=SA" -p 1433:1433 -d --name=sql mcr.microsoft.com/azure-sql-edge

**Comment by dancosta-fed** [comment]
> https://stackoverflow.com/questions/65398641/docker-connect-sql-server-container-non-zero-code-1/66919852#66919852

Thank you! You're a life saver

**Comment by alexandrenolla** [comment]
> I have to say I am surprise that a company like microsoft can't handle that :s Sad to read the thread.

WELCOME TO MICROSOFT.

**Comment by dotnetemmanuel** [comment]
Would this work on Ubuntu on RPI Arm64?

**Comment by justincolangelobrs** [comment]
I am seeing in Docker Desktop version 4.26.1 that the setting is gone. It seems you get it for free now. 

> Improved x86 performance and fixed most common Rosetta issues

**Comment by Mathew-Estafanous** [comment]
I'm faced with a 301 Service Unavailable with any attempt I make to pull sql server.

cmd(s) I've tried:
```
docker pull mcr.microsoft.com/mssql/server:2022-latest

docker pull --platform linux/amd6 mcr.microsoft.com/mssql/server:2022-latest

docker run -e "ACCEPT_EULA=1" -e "MSSQL_SA_PASSWORD=reallyStrongPwd123" -e "MSSQL_PID=Developer" -e "MSSQL_USER=SA" -p 1433:1433 -d --name=sql mcr.microsoft.com/azure-sql-edge
```

Response error:
```
docker: Error response from daemon: received unexpected HTTP status: 503 Service Unavailable.
```

Docker version:
```
Client:
 Cloud integration: v1.0.35+desktop.5
 Version:           24.0.7
 API version:       1.43
 Go version:        go1.20.10
 Git commit:        afdd53b
 Built:             Thu Oct 26 09:04:20 2023
 OS/Arch:           darwin/arm64
 Context:           desktop-linux

Server: Docker Desktop 4.26.1 (131620)
 Engine:
  Version:          24.0.7
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.10
  Git commit:       311b9ff
  Built:            Thu Oct 26 09:08:15 2023
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.6.25
  GitCommit:        d8f198a4ed8892c764191ef7b3b06d8a2eeb5c7f
 runc:
  Version:          1.1.10
  GitCommit:        v1.1.10-0-g18a0cb0
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
````

I've made sure rosette emulation is on as well.
![image](https://github.com/microsoft/mssql-docker/assets/56979977/a3646ca1-c1f7-4b7a-bca1-ceae82d57fc1)

Any suggestions as to what I might be doing wrong?

**Comment by StrictLine** [comment]
> I'm faced with a 301 Service Unavailable with any attempt I make to pull sql server.
> 
> cmd(s) I've tried:
> 
> ```
> docker pull mcr.microsoft.com/mssql/server:2022-latest
> 
> docker pull --platform linux/amd6 mcr.microsoft.com/mssql/server:2022-latest
> 
> docker run -e "ACCEPT_EULA=1" -e "MSSQL_SA_PASSWORD=reallyStrongPwd123" -e "MSSQL_PID=Developer" -e "MSSQL_USER=SA" -p 1433:1433 -d --name=sql mcr.microsoft.com/azure-sql-edge
> ```
> 
> Response error:
> 
> ```
> docker: Error response from daemon: received unexpected HTTP status: 503 Service Unavailable.
> ```
> 
> Docker version:
> 
> ```
> Client:
>  Cloud integration: v1.0.35+desktop.5
>  Version:           24.0.7
>  API version:       1.43
>  Go version:        go1.20.10
>  Git commit:        afdd53b
>  Built:             Thu Oct 26 09:04:20 2023
>  OS/Arch:           darwin/arm64
>  Context:           desktop-linux
> 
> Server: Docker Desktop 4.26.1 (131620)
>  Engine:
>   Version:          24.0.7
>   API version:      1.43 (minimum version 1.12)
>   Go version:       go1.20.10
>   Git commit:       311b9ff
>   Built:            Thu Oct 26 09:08:15 2023
>   OS/Arch:          linux/arm64
>   Experimental:     false
>  containerd:
>   Version:          1.6.25
>   GitCommit:        d8f198a4ed8892c764191ef7b3b06d8a2eeb5c7f
>  runc:
>   Version:          1.1.10
>   GitCommit:        v1.1.10-0-g18a0cb0
>  docker-init:
>   Version:          0.19.0
>   GitCommit:        de40ad0
> ```
> 
> I've made sure rosette emulation is on as well. ![image](https://private-user-images.githubusercontent.com/56979977/298352513-a3646ca1-c1f7-4b7a-bca1-ceae82d57fc1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM3NTgyMjgsIm5iZiI6MTcyMzc1NzkyOCwicGF0aCI6Ii81Njk3OTk3Ny8yOTgzNTI1MTMtYTM2NDZjYTEtYzFmNy00YjdhLWJjYTEtY2VhZTgyZDU3ZmMxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE1VDIxMzg0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE4ODA3OTQwNDk1MzdhZmU0M2RjMWNiMGU5NzY3YTBmZDFlNDc4YWQyMDMyOWZhMmZkOWE0OTg5M2VmNTI0NzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.nX1kuLlYf1jFrtZidb9LitpNijwMcrJRPtVBroINZOc)
> 
> Any suggestions as to what I might be doing wrong?

Your screenshot about the forgotten Rosetta setting saved my day, thanks!

My command (Mac Studio with M2 Max): 
`docker run --name "mssqlex03" --platform=linux/amd64 --hostname "mssqlex03" -e "ACCEPT_EULA=Y" -e "MSSQL_SA_PASSWORD=strongPassword" -e "MSSQL_PID=Express" -p 1433:1433 -d mcr.microsoft.com/mssql/server:2022-CU13-ubuntu-22.04`

**Comment by ddimaria** [comment]
The above didn't work for me.  I ended up installing `mcr.microsoft.com/azure-sql-edge` and using the `go-sqlmd` setup:
```dockerfile
RUN curl -fsSL https://packages.microsoft.com/keys/microsoft.asc | gpg --dearmor -o /usr/share/keyrings/microsoft-prod.gpg
RUN curl https://packages.microsoft.com/config/debian/11/prod.list | tee /etc/apt/sources.list.d/mssql-release.list
RUN apt-get update && ACCEPT_EULA=Y apt-get install -y sqlcmd && apt-get clean && rm -rf /var/lib/apt/lists/*
```

**Comment by StrictLine** [comment]
> The above didn't work for me. I ended up installing `mcr.microsoft.com/azure-sql-edge` and using the `go-sqlmd` setup:
> 
> RUN curl -fsSL https://packages.microsoft.com/keys/microsoft.asc | gpg --dearmor -o /usr/share/keyrings/microsoft-prod.gpg
> RUN curl https://packages.microsoft.com/config/debian/11/prod.list | tee /etc/apt/sources.list.d/mssql-release.list
> RUN apt-get update && ACCEPT_EULA=Y apt-get install -y sqlcmd && apt-get clean && rm -rf /var/lib/apt/lists/*

Thanks for sharing your approach! I'm not quite sure, if I got it right:

- You add the Microsoft package source to the apt package manager and update it.
- You install the sqlcmd --> This is actually the SQL client, how does this suppose to help on the memory mapping (or such) issue of SQL server boot process on Apple Silicon (ARM processor arch.)?

I appreciate any hint!

**Comment by ddimaria** [comment]
> > The above didn't work for me. I ended up installing `mcr.microsoft.com/azure-sql-edge` and using the `go-sqlmd` setup:
> > RUN curl -fsSL https://packages.microsoft.com/keys/microsoft.asc | gpg --dearmor -o /usr/share/keyrings/microsoft-prod.gpg
> > RUN curl https://packages.microsoft.com/config/debian/11/prod.list | tee /etc/apt/sources.list.d/mssql-release.list
> > RUN apt-get update && ACCEPT_EULA=Y apt-get install -y sqlcmd && apt-get clean && rm -rf /var/lib/apt/lists/*
> 
> Thanks for sharing your approach! I'm not quite sure, if I got it right:
> 
> * You add the Microsoft package source to the apt package manager and update it.
> * You install the sqlcmd --> This is actually the SQL client, how does this suppose to help on the memory mapping (or such) issue of SQL server boot process on Apple Silicon (ARM processor arch.)?
> 
> I appreciate any hint!

@StrictLine  Using the go version of `sqlcmd` allowed me to do things like setup users and databases.  Here a link to the Dockerfile I that worked for my Mac (M1ProMax) and in CI in Linux: https://github.com/quadratichq/quadratic/pull/2702/files#diff-2321d37ad093dcff7d29e4f5f0520d9819e0c772e05a2c4d1aeb95f3d76470c2

**Comment by StrictLine** [comment]
Thanks a lot for sharing it, @ddimaria! I'll check it out soon.

**Comment by Meligy** [comment]
I've been away from this for a while. Out of curiosity, is there a way to run MS SQL Full Text Search on Docker for ARM Macs nowadays?

**Comment by amitkh-msft** [comment]
The latest SQL Server 2025 preview images should work on MAC, please let me know if you still see this issue.

@Meligy  regarding FT search on ARM MACs that is not supported as the FT package that is released for Linux is x64 and is not part of the mssql-server container image, thus you may not be able to run this on MAC. I can check and confirm

**Comment by thisisthekap** [comment]
@amitkh-msft So this issue is fixed for SQL Server 2025 preview? (apart from fulltext search)

**Comment by bielu** [comment]
@amitkh-msft i am still seeing issue on arm based cpu, without emulation. I tried it on windows arm but yeah same issue either way.
Just to confirm i pulled image 	2025-latest with sha256:e2e5bcfe395924ff49694542191d3aefe86b6b3bd6c024f9ea01bf5a8856c56e
